redid the YOLO object detection algorithm

TheHidden1 opened the pull request:
redid the YOLO object detection algorithm
more simple & understandable 
no more memory leaks 
performance boost (i think )

a normal usage example should look like this 
```javascript
let options = { url: '...', }
let yolo = new ml5.YOLO(options); 
let loaded = await yolo.loadModel()
if(loaded){
   let results = await yolo.detect(image ||  video || canvas)
}
```
here's a working example(image) ~~https://codepen.io/hiddentn/pen/NBjPKR~~
another one with a video element ~~https://codepen.io/hiddentn/pen/jpmbRy~~

mw108 commented:
`./YOLO/index.js` Line 175:

    const boxClassProbMask = tf.greaterEqual(boxScores, ...);

Isn't this supposed to be `tf.greaterEqual(boxScores1, ...);`?

mw108 commented:
p5 images and video are not supported anymore?

TheHidden1 commented:
@mw108 thanks you  , i guess i missed those mistakes  :) 

concerning p5 images & video i am not sure how to handle them , anyone  is welcome  to  join in

TheHidden1 commented:
here's a working example https://codepen.io/hiddentn/pen/NBjPKR

mw108 commented:
@TheHidden1 Thanks for the quick fixes. :)

I fiddled around a bit with p5 and it turned out that supporting p5 images and webcam is actually pretty easy.

For images, you need to pass `img.canvas` to the `yolo.detec()` function.
Example: https://codepen.io/mw-108/pen/gjWGWX

And for webcam, you need to pass `video.elt` to the `yolo.detect()` function.
Example: https://codepen.io/mw-108/pen/XBRaQG

cvalenzuela commented:
Thanks @TheHidden1 and @mw108! this is great work!
We will need to update way the method is created. This has changed since you forked the library. See: https://github.com/ml5js/ml5-library/blob/master/src/YOLO/index.js#L146

TheHidden1 commented:
1 - redid some pre/post processing stuff 
2 - i need some help with  bench-marking and  testing. I will post soon some updated examples
3 - redid the yolo model : 
  * removed batch norm layers as they are not needed for inference ( batch folding) [here](https://github.com/ModelDepot/tfjs-yolo-tiny/issues/6#issuecomment-388422065)
  * rewrote the yolo model.json  file , it's now more human friendly (removed some of the redundant/auto initialized stuff )
 ml5js/ml5-data-and-models#31

TheHidden1 commented:
@cvalenzuela  i am trying to simplify the post processing  as much as i can & i have some suggestion : 
- do you think that we should output  the bounding box coordinates as is 
 (XCenter,YCenter,Width,Height) instead of the current x1,y1,x2,y2.
my reasoning for this is that i am trying to remove as much operations on the gpu back-end as i can
see  this https://github.com/ModelDepot/tfjs-yolo-tiny/issues/6.
- i am working atm on yolov3-tiny and the full yolov3 model , i managed to get them working but i am still having some issues the post-processing side of things (the bounding box coordinates) how do you  think they should be implemented ?  does it make sense to separate  each model : 
```javascript
const yolov2 = new ml5.yolov2(options)
const yolov3 = new ml5.yolov3(options)
```
or 
```javascript
const yolo = new ml5.yolo(flavor,options)
 ```
keep in mind that there are some differences in both models most importantly : 
   * there are some post-processing differences / mainly yolov3-tiny  has 2 ouput tensors : 13x13 & and an  up-sampled output for smaller  objects (26x26) (3 output in the case of the full yolo v3) .
   * there are some differences  in how the models were made : v2 uses a sequential model whereas v3 smaller ones with some concatenation .

shiffman commented:
For integration with p5 receiving the bounding box as `x,y,w,h` is probably ideal since those are the defaults with `rect()`. Thanks for your contributions!

cvalenzuela commented:
- Following @shiffman, `x,y,w,h` makes sense!

- For managing different YOLO versions, I think we can support two cases:
```javascript
const yolo = new ml5.yolo('v2')
```
and also this:
```javascript
const options = { 
  version: '3'
 // other options
} 
const yolo = new ml5.yolo(options)
```

shiffman commented:
Aligning with our A PI though we would wrap in a function (avoiding `new`) like so? And `YOLO` all caps?

```javascript
const yolo = ml5.YOLO('v2')
```

cvalenzuela commented:
ups! yes, missed that!

TheHidden1 commented:
i think yolov2 is ready for a merge , it output  [x y w h] now.(x,y being to coordinates for the center no top)
am holding off yolov3 atm as it's not ready yet / logically the post processing should be the same as yolov2 but i am finding some difficulties with the  box coordinates  
anyway v3 is not that different than v2 in terms of performance / accuracy 
tiny-yolov2:
![v2-test2](https://user-images.githubusercontent.com/31354864/44142354-37433568-a080-11e8-927e-956642d27065.png)
tiny-yolov3:
![v3-test2](https://user-images.githubusercontent.com/31354864/44142362-3d227124-a080-11e8-91c4-198bc4f636c0.png)

cvalenzuela commented on src/YOLO/index_test.js:
```diff
@@ -25,14 +25,15 @@ describe('YOLO', () => {
 
```
The test error is:
```javascript
TypeError: Cannot read property 'filterBoxesThreshold' of undefined
```
can you try defining the variable here first. Like this: https://github.com/ml5js/ml5-library/blob/master/src/ImageClassifier/index_test.js#L20

cvalenzuela commented:
@TheHidden1, let me know if you want us to continue to merge this!

TheHidden1 commented:
Sorry for the late reply , i had a hard drive failure and i lost all my stuff ,  you fix the test error and merge if you want . I 2 commits localy that i can push as soon as i fix things up

cvalenzuela commented:
sounds good. we can wait for your commits and we cant then fix the tests

TheHidden1 commented:
i think you  can push now

joeyklee commented:
Hello @TheHidden1 
Thanks for all your work on this PR. As the ml5 crew going back through the existing PRs after the busy last months, we're curious to know whether or not you're still interested in integrating refactor of the YOLO for ml5. If so, would you be able to update your PR and check for any breaking changes or issues? 

I've tried to run your re-implementation and I get the following error at the `.predict()` function in YOLO. It seems that there's an issue with the incoming image data and for whatever reason the `.predict()` function isn't returning anything.

<img width="639" alt="screen shot 2019-01-24 at 12 47 08" src="https://user-images.githubusercontent.com/3622055/51698405-178ca100-1fd8-11e9-9393-f17df24f681a.png">
<img width="513" alt="screen shot 2019-01-24 at 12 47 19" src="https://user-images.githubusercontent.com/3622055/51698406-178ca100-1fd8-11e9-91d7-2c75571d21eb.png">

If I run the YOLO detection on a simple example as it is currently implemented in ml5.js v0.1.3, then this is what I see:

<img width="889" alt="screen shot 2019-01-24 at 13 00 12" src="https://user-images.githubusercontent.com/3622055/51698407-178ca100-1fd8-11e9-9a46-db186b9eddb6.png">

Let us know if you're interested to explore this further. Many thanks!

TheHidden1 commented:
i will check it out right away

joeyklee commented:
@TheHidden1 - Thanks so much for the updates! @shiffman and I will have a look this week and let you know if/when we merge! Many thanks!

TheHidden1 commented:
@shiffman @joeyklee i cleaned up some stuff and added support for yolo v3 models :
* one major thing is that  i added is the ability to change the model input size witch gives a good trade off between accuracy and speed . please see ml5js/ml5-data-and-models#31


![g](https://user-images.githubusercontent.com/31354864/51986520-af8afe80-24a0-11e9-9333-315405b60efa.PNG)

* I restructured  the post processing algorithm to be more clear & efficient (i think) , there is also a V2
that implements `tf.image.nonMaxSuppression()` but it seems to be  slowing  things down a bit (10 - 20 ms) 

* this how the new config shoud be : 
```javascript
     // this an example for the tiny yolov2 model
 let config = {
     version: 'v2', /* 'v2' || 'v3' */

     /*
       128 || 144 || 224 || 256 || 320 || 416 (or any multiple of 32 rly)
       we can change this on the fly  now wohoo!
      */
     modelSize: 416 ,   

     URL: '',

     /* inference parameters */
     IOUThreshold: 0.5,
     classProbThreshold: 0.4,

     /* this mask defines  witch anchors go to witch layer 
         eg : for tiny yolo v3 :  ( hast 2 output layers at different scales ) 
                masks: [ [3, 4, 5], [0, 1, 2] ],
                anchors: [[10, 14], [23, 27], [37, 58], [81, 82], [135, 169], [344, 319] ],
     */
     masks: [ [0, 1, 2, 3, 4] ],
     anchors: [ [0.57273, 0.677385], [1.87446, 2.06253], [3.33843, 5.47434], [7.88282, 3.52778], [9.77052, 9.16828]],

     /* class names array  */
     classes: CLASS_NAMES_COCO,
}
```
* i did some benchmarks  & here are some of the results  : 
    * Chrome : 71.0.3578.98 (Official Build) (64-bit)
    * CPU : Intel® Core™ i7-7700HQ Processor 
    * GPU0 : Intel® HD Graphics 630
    * GPU1 : NVIDIA GeForce GTX 1050 Ti
    * Backend :  webgl

<table>
    <tr>
        <th colspan="7">ml5.js + tfjs 0.13</th>
    </tr>
    <tr> 
        <th colspan="1"></th>
        <th colspan="3">CPU*</th>
        <th colspan="3">GPU**</th>
    </tr>
    <tr>
        <th colspan="1"></th>
        <th colspan="1">Min</th>
        <th colspan="1">Avg</th>
        <th colspan="1">Max</th>
        <th colspan="1">Min</th>
        <th colspan="1">Avg</th>
        <th colspan="1">Max</th>
    </tr>
    <tr>
        <td colspan="1">416x416</td>
        <td colspan="1">522.99</td>
        <td colspan="1">538.13</td>
        <td colspan="1">556.40</td>
        <td colspan="1">98.89</td>
        <td colspan="1">101.68</td>
        <td colspan="1">114.79</td>
    </tr>
    <tr>
        <td colspan="1">320x320</td>
        <td colspan="1">325.30</td>
        <td colspan="1">332.64</td>
        <td colspan="1">351.39</td>
        <td colspan="1">58.69</td>
        <td colspan="1">61.40</td>
        <td colspan="1">71.79</td>
    </tr>
    <tr>
        <td colspan="1">224x224</td>
        <td colspan="1">186.10</td>
        <td colspan="1">191.98</td>
        <td colspan="1">207.30</td>
        <td colspan="1">33.90</td>
        <td colspan="1">36.58</td>
        <td colspan="1">89.29</td>
    </tr>
    <tr>
        <td colspan="1">128</td>
        <td colspan="1">83.55</td>
        <td colspan="1">78.39</td>
        <td colspan="1">104.00</td>
        <td colspan="1">15.60</td>
        <td colspan="1">16.95</td>
        <td colspan="1">23.90</td>
    </tr>
    <tr>
        <th colspan="7">tfjs 0.14.2</th>
    </tr>
    <tr>
        <th colspan="1"></th>
        <th colspan="3">CPU*</th>
        <th colspan="3">GPU**</th>
    </tr>
    <tr>
        <th colspan="1"></th>
        <th colspan="1">Min</th>
        <th colspan="1">Avg</th>
        <th colspan="1">Max</th>
        <th colspan="1">Min</th>
        <th colspan="1">Avg</th>
        <th colspan="1">Max</th>
    </tr>
    <tr>
        <td colspan="1">416x416</td>
        <td colspan="1">464.39</td>
        <td colspan="1">476.80</td>
        <td colspan="1">525.49</td>
        <td colspan="1">83.30</td>
        <td colspan="1">87.24</td>
        <td colspan="1">126.70</td>
    </tr>
    <tr>
        <td colspan="1">320x320</td>
        <td colspan="1">275.59</td>
        <td colspan="1">283.80</td>
        <td colspan="1">296.49</td>
        <td colspan="1">54.99</td>
        <td colspan="1">56.86</td>
        <td colspan="1">67.00</td>
    </tr>
    <tr>
        <td colspan="1">224x224</td>
        <td colspan="1">140.79</td>
        <td colspan="1">144.96</td>
        <td colspan="1">150.40</td>
        <td colspan="1">33.89</td>
        <td colspan="1">36.27</td>
        <td colspan="1">45.40</td>
    </tr>
    <tr>
        <td colspan="1">128x128</td>
        <td colspan="1">76.89</td>
        <td colspan="1">79.98</td>
        <td colspan="1">92.90</td>
        <td colspan="1">20.20</td>
        <td colspan="1">22.40</td>
        <td colspan="1">32.00</td>
    </tr>
</table>


###### * Intel® HD Graphics 630
###### ** NVIDIA GeForce GTX 1050 Ti


here is an updated demo where you can try out all of this stuff  : https://codepen.io/hiddentn/full/gqrBeO

note : there is a big time difference when using `detect()`  vs  `detectAsync()` (+100ms in inference time) ,so please can anyone test things out ( there is a detect & a detect Sync buttons in the DEMO) and report their findings thank you

joeyklee commented:
Hi @TheHidden1 Wow! This is wild. I didn't manage to have a deep look at your contributions yet, but I hope to do some checks with @shiffman soon. I just wanted to check in to let you know it's still very much on our minds! Thanks!

joeyklee commented:
Hi @TheHidden1 - apologies for the radio silence. @yining1023 and I had a deeper look into your code and we have the feeling that your current proposal diverges a bit too far from the current API structure. We recognize that we don't have good (or any?) documentation yet about the API conventions for ml5.js but that is high on our list. 

We would like to propose the following:

1. Please submit a simple working example of your refactor in action, preferably in p5.js and make a PR to the ml5-examples repo. 
2.  If you're still interested to work on adapting your proposal closer to the current API structure of YOLO, you're welcome to do this and we can revisit this. 

Thanks again for all your work on this! cc/ @shiffman

TheHidden1 commented:
i try & do my best

TheHidden1 commented:
i add a small demo https://github.com/ml5js/ml5-examples/pull/107

TheHidden1 commented:
you know what let me start again .this has become too messy
